### ISSUE TITLE
Make Jnario a pure Xbase language

### ISSUE THREAD

**Issue by borisbrodski** [sebastianbenz/Jnario#160]
### Motivation

Jnario currently depends directly on Xtend. Since Xtend provides no official public stable API, even each minor release of Xtend may break Jnario and this already happened in the past.

The suggestion is to make Jnario pure Xbase language dropping dependency on Xtend.

Pros:
- Reliable dependency management (dependencies like `[2.6.0,2.7.0)` will not break Jnario)
- Less risk for users to end up with a broken installation
- Easy and quick upgrade to new major versions of Xtext

Cons:
- Drop support of the features
  - Class definitions inside specs
  - Anonymous classes
  - Template expressions (partially)
- Need to reimplement or copy some logic from Xtend
### TOP PRIORITY
- [x] Fix outline (prevent outline update on any change, remove Xbase expression nodes)
- [x] Fix jump to spec/feature from JUnit view (double click) (@SebastianPoetzsch pls, take a look)?
- [ ] Test standalone batch compiler
### Assignments
- Boris
  - [ ] Fix Formatter tests (writing basic formatters for spec & feature)
  - [ ] Test standalone batch compiler
#### Plan
- Common
  - [x] Fix tests: org.jnario.lib.tests
  - [x] Fix tests: org.jnario.standalone.tests
  - [x] Fix tests: org.jnario.tests
  - [x] Fix tests: org.jnario.ui.tests
  - [x] Migrate to the latest Xtext
  - [ ] Check GUI: preferences
  - [x] Fix Maven build without running test (`mvn clean install -DskipTests`)
  - [x] Fix Maven complete build (`mvn clean install`)
  - [ ] Test Jnario in production environments
  - [ ] Migrate to Java 8 (when Xtext does)
- Spec
  - [x] Regenerate EMF and Xtext code (xtend-free)
  - [x] Comment out as much as possible (adding `// TODO NO_EXTEND` marker)
  - [x] Get everything else compile
  - [x] Get everything else run
  - [ ] Uncomment `TODO NO_EXTEND` parts and get it to work
  - [ ] Rewrite Formatter
  - [ ] Check UI (outline, autocompletion, quick fixes, ...)
- Feature
  - [x] Regenerate EMF and Xtext code (xtend-free)
  - [x] Comment out as much as possible (adding `// TODO NO_EXTEND` marker)
  - [x] Get everything else compile
  - [x] Get everything else run
  - [ ] Uncomment `TODO NO_EXTEND` parts and get it to work
  - [ ] Rewrite Formatter
  - [ ] Check UI (outline, autocompletion, quick fixes, ...)
- Suite
  - [x] Regenerate EMF and Xtext code (xtend-free)
  - [x] Comment out as much as possible (adding `// TODO NO_EXTEND` marker)
  - [x] Get everything else compile
  - [x] Get everything else run
  - [ ] Uncomment `TODO NO_EXTEND` parts and get it to work
  - [ ] Rewrite Formatter
  - [ ] Check UI (outline, autocompletion, quick fixes, ...)
## Infos

Progress:
- https://github.com/borisbrodski/Jnario/tree/no_xtend
- https://github.com/borisbrodski/Jnario/tree/no_xtend_xtext2.9

**Comment by borisbrodski** [comment]
Migrated "no_xtend" branch to Xtext 2.8.1

**Comment by sebastianbenz** [comment]
This is fantastic! How is your progress?

**Comment by borisbrodski** [comment]
Pretty good I think. Check it out: https://github.com/borisbrodski/Jnario/tree/no_xtend

With Xtext 2.8.1 (and no dependencies to Xtend) I'm able to compile and run very simple feature, very simple spec and a corresponding suite. :)

**Comment by sebastianbenz** [comment]
Great! Do think that it will be possible to port all features to your
version?

**Comment by borisbrodski** [comment]
I'm not sure about all features, but most features with no doubt.
Currently I'm going through the tests to better understand the situation.

**Comment by borisbrodski** [comment]
By the way, could you please help me to remove dependency to xtend-maven-plugin from the jnario-maven-plugin project?

I'm not that experienced in maven plugins and it's not a requirement for my project here.

**Comment by sebastianbenz** [comment]
Are you sure that the jnario maven plugin has a dependency to the xtend
plugin? I think I duplicated the code anyway.

**Comment by borisbrodski** [comment]
Yes, at least

```
import org.eclipse.xtend.core.XtendStandaloneSetupGenerated;
import org.eclipse.xtend.core.xtend.XtendPackage;
import org.eclipse.xtend.core.XtendRuntimeModule;
import org.eclipse.xtend.core.compiler.batch.XtendBatchCompiler;
```

Why do you need all this maven stuff any way? Xbase projects can be build from maven without any coding needed. Did you added some cool features to the maven build?

**Comment by sebastianbenz** [comment]
I think I copied the corresponding Xtend classes into the jnario plugin. I
didn't add any new features, the only difference was the way the
ResourceSet is loaded. However, I think all these moved into the
JnarioBatchCompiler anyway.

**Comment by borisbrodski** [comment]
Do you have an idea, why I can't resolve references (by name) to specs and features from suite? (\.*\ works)

In editor (during typing) references are resolved and autocompletion works as expected. But when I save the suite referencing a feature or a spec it generates an error "can't resolve reference to ...".

One of standalone compiler tests shows the same error. 

Thank you!

**Comment by sebastianbenz** [comment]
I remember this being quite tricky, but I only had this problem with the standalone compiler.

**Comment by borisbrodski** [comment]
The error `Couldn't resolve reference to Specification 'Example'` occurs only if suite-resource get added to the ResourceSet prior to the spec-resource. If this is the case and the suite get loaded first, the spec "Example" isn't yet loaded and can't be resolved. The error message is now stored within the suite-resource.

After spec resource and all other resources are loaded the XtendBatchCompiler tries to redo resolving:

```java
// install a fresh type provider for the second phase, so we
// clear all previously cached classes and misses.
installJvmTypeProvider(resourceSet, classDirectory, false);
EcoreUtil.resolveAll(resourceSet);
```

But this doesn't work. The old error message stays within the suite-resource, no scope and linking methods get called.

The error must be somewhere outside of the standalone-compiler stuff, since almost exactly the same happens within a runtime eclipse.

Some caching issues? Any ideas?
@oehme What would you do to find the problem?

Thank you!

**Comment by oehme** [comment]
Sounds like cross references are resolved during indexing, which is
generally not possible. The most likely candidates are the inferer and
resourcedescriptionstrategy.
Am 08.04.2015 13:31 schrieb "Boris Brodski" <notifications@github.com>:

**Comment by borisbrodski** [comment]
Got it! Thank you very much!! This was the `Linker` instead of `LazyLinker` :/

**Comment by borisbrodski** [comment]
A lot of Jnario tests looks like this one:

```java
	fact "returns steps"{
		'''
		Feature: My feature
		Scenario: My first Scenario
			Given something
		'''.parseScenario
		scenario("Scenario: My first Scenario").executables => list(step("Given something"))
	}
```

Since we don't have the rich string support yet, I see following four possibilities:

- Replace all `'''` => `'`
- Add new string type `'''` ... `'''` (very simple)
- Add rich string support with placeholders but without IF and FOR (moderate complexity ?)
- Add full support for rich strings

Any ideas?

**Comment by sebastianbenz** [comment]
I would go with option two (Add new string type ''' ... ''). Having proper
multi-line strings is very important - however I seldom use actual rich
strings in my specs. This way you also have the possibility to add proper
rich string support later.

**Comment by oehme** [comment]
Please note that the default string literals ("...") in Xbase are already
multiline.
Am 10.04.2015 13:01 schrieb "Sebastian Benz" <notifications@github.com>:

**Comment by sebastianbenz** [comment]
I still would prefer having '''...''' as well. You no longer need to care
about quotes and double quotes in your text and it makes it easier to
migrate tests.

**Comment by oehme** [comment]
Yes, you can do that by overriding the STRING terminal rule and its value
converter. No need for the huge amount of code related to template
expressions. =)

Also, while you're talking about Strings: You can overwrite
XbaseTypeComputer#_computeTypes(XStringLiteral) so that a single character
in single quotes is always a char and never a String. We had to do that
conversion for backwards compatibility, but it causes major headaches for
newcomers, especially when using assertEquals/shouldBe.

2015-04-10 14:16 GMT+02:00 Sebastian Benz <notifications@github.com>:

**Comment by borisbrodski** [comment]
Is there any particular reason why you create jvm object through `TypesFactory` like here 

```java
def toClass(JnarioClass xtendClass, List<JvmGenericType> scenarios, IJvmDeclaredTypeAcceptor acceptor, List<Runnable> doLater, boolean preIndexingPhase){
	val javaType = typesFactory.createJvmGenericType
	setNameAndAssociate(xtendClass.jnarioFile, xtendClass, javaType)
	acceptor.accept(javaType)
	if (!preIndexingPhase) {
		doLater.add([|init(xtendClass, javaType, scenarios)]);
	}
	javaType
}
```

Why not just to use XBase `eObject.toClass(...)` approach?

PS
The `toClass` method above produces an error `... cannot be exported as the target is not contained in a resource`. Why did it worked earlier?

**Comment by sebastianbenz** [comment]
The scenarios are passed in as well. I think I needed them later in the
process. The implementation should be the same as in xbase. Did it change
in 2.8?

**Comment by borisbrodski** [comment]
In your Jnario tests you use pretty often the construct

```
  /**
   * Docs
   * Docs
   * Docs
   */
  fact "description" {
    "not implemented"
  }
```

Unfortunately this produces now (with Xbase 2.8.1) the error 

    This expression is not allowed in this context, since it doesn't cause any side effects.

How should I deal with it:

- Convert to `// not implemented` (this will make test `@Ignored`)
- Convert to ` "not implemented" should not be null`
- Something else

**Comment by sebastianbenz** [comment]
The reason why I add the string is that the spec is not marked as pending -
to fix the formatting of the spec. When I think about it, the correct
solution would be to only format pending specs differently if we generate
the docs after an actual test run (including failed and passed specs). This
way we could simply convert this into a normal comment.

**Comment by borisbrodski** [comment]
Ok, so I just remove `"not implemented"` strings for now and we will fix the formatting later on. Ok?

**Comment by sebastianbenz** [comment]
SGTM

**Comment by riederm** [comment]
Hi guys,
does anybody have something like an updated list of open TODOs. @ghaith me are planning to join the efforts.
So from this thread and the current state of the noXtend branch I can tell that following things are still todo(?)
* Formatters (maybe redo them as a whole and use xtext's new formatter-api?)
* Maven Plugin
* Rich Strings
* Fix tests

**Comment by borisbrodski** [comment]
Hi everyone,

back from the vacation I continue my work on Jnario. This is the only reasonable way to migrate my customer projects to the new Xtext.

At first I will update the TODO list at the top of this page adding elements from @riederm and some new ones. (How can I give permissions to others to check/uncheck checkboxes in the TODO list?)

For now I will concentrate on getting Jnario tests green. It would be really great, if someone could get the maven build to work. Just get `mvn clean install -DskipTests` to work.

I'm pretty sure, together we will release first usable Xtend-only version in the near future. Thank you!

**Comment by riederm** [comment]
hi boris, great news, same situation for me, we're stuck at xtext 2.6.3  until we're successful

i tried to fix some minor bugs, nothing spectecular, but I found myself most of the time in a situation that tests would pass, if the rich strings used in the facts would work as expected. I spent a night trying to extract the rich-string classes from xtend. I stopped when I realized that the xtend rich-strings have a lot of features (like the <<IF...>> stuff, the <<FOR...>> stuff, special highlighting provider, etc.) ... it was too much for that particular night :-)

The maven clean build runs for me except for the Jnario Maven Plugin which still uses some XtendCompilerXXX classes. I'll take a look at these next week.

Keep me up to date if you start some work on the rich strings, I'm interested in seeing your approach.

**Comment by oehme** [comment]
You would save yourselves a lot of trouble if you dropped rich strings.
They are about 5000 lines of code that you would have to maintain.

I looked at a few tests and most of them have a RichStrings without any
replacement variables. You can replace those with normal Strings. Normal
Strings in Xbase are multi-line-capable.

2015-08-19 20:30 GMT+02:00 Mathias Rieder <notifications@github.com>:

**Comment by riederm** [comment]
basically I agree, the tests failed mostly because of different indentation
of the resulting strings.
I also gravitate towards the idea of using the Xbase-Strings and enhance
them to preserve the indentation in multiline strings

for a short term solution i would be totally happy with that, midterm I
guess I would need to refactor a lot of tests in our existing projects to
avoid the replacement variables. I guess it would be the same for boris?
its a nice feature though :-)

**Comment by oehme** [comment]
You can combine Xtend and Jnario. Have an Xtend class with the templates
and include it as an extension.

Honestly, I would suggest some even more radical simplifications. Getting
rid of method declarations for instance. If you have arbitrary methods and
inheritance, you need to have validation, auto completion and quick fixes
for overloading and overriding. These used to come from the Xtend classes
that you inherited. You would have to reimplement all that.



2015-08-19 21:08 GMT+02:00 Mathias Rieder <notifications@github.com>:

**Comment by borisbrodski** [comment]
Hi, my colleague had implemented rich strings indentation and will push it very soon. This will fix a lot of tests. Also the @Rule extension fields in specs were broken (generated default visibility instead of public). I will push the fix today.

@oehme in our test base we have a lot of methods within specs. Extracting all those methods will not only mean a lot of work, but will also reduce readability of the tests. I will suggest to keep simple method definitions without any extra support for advance features (for now). Later on we can improve on this.

**Comment by borisbrodski** [comment]
Just pushed some new commits including new implementation of the rich string indentation.

Tests run: 458, Failures: 21, Errors: 10, Skipped: 17

**Comment by borisbrodski** [comment]
Hi,

we are stuck with the following problem: In the second eclipse instance with the Jnario test projects opened we get a bunch of errors (NullPointerExceptions), that we tracked down to the failing `JnarioNameProvider`.

Here is the outline of the events

* `org.jnario.standalone.tests/testdata/test/Example.suite` is located in a simple folder, but still wants to be compiled (Xtend skips such files)
* `Example.suite` references `SuiteScopeProvider.spec` using `\.*\` pattern
* Jnario trys to determine the java class name for the spec calling `ExampleNameProvider.resolveProxyTypeName()` method
* `ExampleNameProvider.resolveProxyTypeName()` method uses the node model to get the name of the class without resolving all the proxies.
* It calls `NodeModelUtil.findNodesForFeature()` that calls `NodeModelUtil.getNode()`. Both methods return `null`.
* The inspection reveals, that the corresponding EObject `ExampleGroupImplCustom` contains no `ICompositeNode` adapter whatsoever :(

As a result, the `ExampleNameProvider.resolveProxyTypeName()` returns

* `org.jnario.suite.unit.Spec`
instread of 
* `org.jnario.suite.unit.SuiteScopeProviderSpec`


Any ideas, how we can fix it? (Probably by not compiling files from non-classpath folders)

**Comment by sebastianbenz** [comment]
Not compiling files from non-classpath folders sounds for me like the way to go. Jnario should behave exactly the same way xtend does.

**Comment by borisbrodski** [comment]
Any idea how to ignore .spec, .feature and .suite files, that are not in a source folder?
(Couldn't find, how Xtend does it)

Thank you.

**Comment by riederm** [comment]
I once stumbled across something while debugging inside the XtextBuilder which looked interesting. 
org.eclipse.xtext.builder.impl.ToBeBuiltComputer.isHandled(IStorage)
```java
/**
* Returns <code>true</code> if the storage should be processed by the builder.
* That is, it is known to the {@link UriValidator} and a {@link IToBeBuiltComputerContribution contribution}
* can possibly manage it or it is a file.
* 
* @see UriValidator#isPossiblyManaged(IStorage)
* @see IToBeBuiltComputerContribution#isPossiblyHandled(IStorage)
*/
protected boolean isHandled(IStorage storage) {
	boolean possiblyManaged = contribution.isPossiblyHandled(storage);
	if ((possiblyManaged || storage instanceof IFile) && uriValidator.isPossiblyManaged(storage)) {
		return true;
	}
	return false;
}
```

so maybe rebind to an uriValidator that is aware of the source-folders?

**Comment by borisbrodski** [comment]
Thank you! Found it. It's XtendResourceUiServiceProvider.java.

I copied this class to `org.jnario.ide` package of the `org.jnario` project. I also added `org.eclipse.xtext.ui` dependency to the project. Is it OK?

I think, it would be better to have something like `org.jnario.ui` project and add the new class there.

PS
All 4 test projects are green. The fix pushed.

**Comment by borisbrodski** [comment]
Another problem, that stays in a way a little bit is, that I can't run mockito based tests directory from Eclipse (have to run it through mvn).

The error I get is: 

<pre>
Caused by: java.lang.SecurityException: class "org.jnario.report.SpecExecutionAcceptor$$EnhancerByMockitoWithCGLIB$$3393b8bf_3"'s signer information does not match signer information of other classes in the same package
	at java.lang.ClassLoader.checkCerts(ClassLoader.java:806)
	at java.lang.ClassLoader.preDefineClass(ClassLoader.java:487)
	at java.lang.ClassLoader.defineClassCond(ClassLoader.java:625)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:615)
	... 71 more
</pre>

(SpecResultParser)

Any experience fighting this beast?

**Comment by sebastianbenz** [comment]
I wasn't able to run mockito tests in the eclipse runtime as well. I
implemented my tests in a way that I could run them without eclipse (at
least the non-ui tests).

**Comment by gzunino** [comment]
I have this problem before, and solved by using mockito from https://github.com/jbosstools/jbosstools-locus

**Comment by borisbrodski** [comment]
Check out my no_xtend branch
All tests are green!! :)))

https://github.com/borisbrodski/Jnario/tree/no_xtend

<pre>
$ mvn clean install
...
...
...
[INFO] Reactor Summary:
[INFO]
[INFO] jnario ............................................ SUCCESS [0.799s]
[INFO] jnario Target Definition .......................... SUCCESS [0.185s]
[INFO] org.jnario.lib .................................... SUCCESS [2.617s]
[INFO] org.jnario ........................................ SUCCESS [5.432s]
[INFO] org.jnario.feature ................................ SUCCESS [4.701s]
[INFO] org.jnario.ui ..................................... SUCCESS [1.753s]
[INFO] org.jnario.spec ................................... SUCCESS [6.619s]
[INFO] org.jnario.feature.ui ............................. SUCCESS [21.408s]
[INFO] org.jnario.spec.ui ................................ SUCCESS [5.318s]
[INFO] org.jnario.suite .................................. SUCCESS [2.117s]
[INFO] org.jnario.suite.ui ............................... SUCCESS [2.182s]
[INFO] org.jnario.standalone ............................. SUCCESS [0.704s]
[INFO] org.jnario.ui ..................................... SUCCESS [0.156s]
[INFO] org.jnario ........................................ SUCCESS [0.174s]
[INFO] org.jnario.m2e .................................... SUCCESS [0.815s]
[INFO] org.jnario.m2e .................................... SUCCESS [0.140s]
[INFO] org.jnario.sdk .................................... SUCCESS [0.158s]
[INFO] org.jnario.updatesite ............................. SUCCESS [5.888s]
[INFO] org.jnario.tests .................................. SUCCESS [1:08.324s]
[INFO] org.jnario.standalone.tests ....................... SUCCESS [21.629s]
[INFO] org.jnario.lib.tests .............................. SUCCESS [9.256s]
[INFO] org.jnario.ui.tests ............................... SUCCESS [35.983s]
[INFO] Jnario Runtime Library ............................ SUCCESS [10.746s]
[INFO] Jnario Standalone Support ......................... SUCCESS [6.697s]
[INFO] Jnario Maven Plugin ............................... SUCCESS [25.709s]
[INFO] Jnario Maven Report Plugin ........................ SUCCESS [4.249s]
[INFO] Create a simple Jnario project .................... SUCCESS [1.652s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 5:28.783s
[INFO] Finished at: Wed Sep 09 17:18:18 CEST 2015
[INFO] Final Memory: 152M/813M
[INFO] ------------------------------------------------------------------------
</pre>

**Comment by sebastianbenz** [comment]
Fantastic! Great work!

**Comment by borisbrodski** [comment]
An upgrade to Xtext 2.9 (latest) seams to be complicated. Lots of classes was removed or renamed, like

* EclipseWorkspaceConfigProvider
* AbstractFileSystemSupport
* EclipseFileSystemSupportImpl
* TraceForStorageProvider
* UnresolvedAnnotationTypeAwareMessageProducer

I know, how to port formatter and JnarioBatchCompiler. But it would be really great, if some one could look through other errors. I will push the new target platform `mars-xtext2.9.target`.

Thank you!

PS
I plan to test Jnario with our "large" project (after the port to the Xtext 2.9 is completed).

**Comment by oehme** [comment]
The FileSystemSupport API is only there for Xtend active annotations, so
that should not be used in Jnario at all.

The tracing API has been redone to factor out the platform independent
parts. I currently don't have time to look into the errors. Did you use the
API or subclass it?

2015-09-25 17:14 GMT+02:00 Boris Brodski <notifications@github.com>:

**Comment by borisbrodski** [comment]
Hi Stefan,

I found only

	public Class<? extends ITraceForStorageProvider> bindTraceInformation() {
		return TraceForStorageProvider.class;
	}

in all Modules. It compiles after reorganizing import.

Also I removed all bindings of FileSystem stuff. Thank you!

**Comment by borisbrodski** [comment]
The new-line problems made me mad. I removed all window-style new line from the project. Rebases and merges over this commit may result in huge merge conflicts, unless `-Xignore-all-space` parameter is specified. Like

    $ git rebase origin/master -Xignore-all-space

Btw @oehme do you know, when Xtext 2.9 will be officially released?

**Comment by oehme** [comment]
The plan is to release it before EclipseCon Europe, i.e. at the end of October.

**Comment by borisbrodski** [comment]
Thanks for the info. The migration to Xtext 2.9 (latest) is almost done.

One thing that bothers me is, that the Xtext 2.9 (latest)  IMHO doesn't have corresponding maven repository. So I can't run the maven build with the new Xtext. Any ideas, what to do with this one?

**Comment by oehme** [comment]
The latest snapshots can always be found at
https://oss.sonatype.org/content/repositories/snapshots

2015-09-30 10:30 GMT+02:00 Boris Brodski <notifications@github.com>:

**Comment by borisbrodski** [comment]
Please, if you want to work on a task, drop me a message, so I can change the assignments and though prevent collisions.

Thank you!

**Comment by borisbrodski** [comment]
@oehme super cool! Thank you!

@sebastianbenz One question: In the grammars you use ValidID all over the place. What's the point of using ParserRule instead of the TerminalRule? Thank you!

From `Xtype.xtext`

    ValidID:
        ID;

**Comment by sebastianbenz** [comment]
Good question - I can't remember the particular reason, but I think Xtend
was using it.

**Comment by oehme** [comment]
Xtend uses that rule to allow some keywords as Valid identifiers. But if it is just "ID", then that rule no longer serves a purpose.

**Comment by ghaith** [comment]
Hello @borisbrodski , 
@riederm,  @chrismathis and me will be trying to run Jnario on our code base to see what is still missing from a "real world project"'s point of view and start tackling these tasks. Our tests code is based 99% on Specs. We use Richstrings extensively as well as lambdas and in some cases methods defined within the spec. So we could probably assume that this is where our focus would be.

**Comment by borisbrodski** [comment]
Hello,

Great! The lambdas and methods defined within specs shouldn't be a problem. The RichString support on the other hand is currently very restricted. You can use neither placeholders nor FOR and IF expressions. The indentation would work as expected though.

So, what do you need to start testing Jnario?

**Comment by borisbrodski** [comment]
@oehme One question: It looks like with Xtext 2.9 scoping in Jnario stops working. Binding with `bindXbaseBatchScopeProvider`, `getScope` get called only for `importedType` and `type` references for a fairly large spec file. Any ideas why?

**Comment by oehme** [comment]
No, nothing that comes to mind. You'll have to debug DefaultLinkingService#getLinkedObjects to see why your ScopeProvider is not called.

**Comment by borisbrodski** [comment]
One question: How can I write a validator for a terminal?
I would like to issue an error, if guillemets are used (that we don't support yet). Thanks.

PS
Just ran Jnario with our big project. Looks great!

**Comment by oehme** [comment]
You can write a ValueConverter and throw a ValueConverterException.

2015-10-05 17:10 GMT+02:00 Boris Brodski <notifications@github.com>:

**Comment by borisbrodski** [comment]
@oehme It looks like the new Formatter API is still buggy. I get a bunch of weird exceptions trying to port the formatter to 2.9.latest. I can't reproduce those exceptions with domain-model-example easily though (to report bugs), but I think, the problem is in Xtext (at least there are some checks with good error messages missing).

Stack traces: https://github.com/borisbrodski/Jnario/issues/5

Could you (or Moritz) just clone the Jnario and take a look of the exceptions ourselves please?
(Before Xtext 2.9 officially released!)

Repo+Branch: https://github.com/borisbrodski/Jnario/tree/no_xtend_xtext2.9

Try to open and format ExampleColumn.spec and you will get exceptions, like in https://github.com/borisbrodski/Jnario/issues/5

Thank you!

**Comment by oehme** [comment]
I forwarded this to Moritz, hopefully he'll be able to take a look.

**Comment by meysholdt** [comment]
ok, I'll try to take a look later this/next week. 

Exceptions thrown from org.eclipse.xtext.formatting2.regionaccess.** are always a bug in Xtext.

**Comment by borisbrodski** [comment]
@meysholdt Thank you! I think, may be the following bug is also connected to the problem: https://bugs.eclipse.org/bugs/show_bug.cgi?id=478711

**Comment by borisbrodski** [comment]
Hi @meysholdt ! Is there a chance, that you will find a minute (or an hour :) ) to fix this formatter bug prior to the 2.9.0 release? Thank you!

**Comment by borisbrodski** [comment]
Come and hear me talking about Jnario at DemoCamp Bonn :)
https://wiki.eclipse.org/Eclipse_DemoCamp_November_2015/Bonn

**Comment by borisbrodski** [comment]
Migrated to Xtext 2.9 (latest)

* `mvn clean install` => `BUILD SUCCESS`
* `mvn org.jnario:report:generate` => `BUILD SUCCESS`

Had to completely rewrite the Maven plugins and maven example.

@sebastianbenz Could you please review the "Complete rework Maven build" commit?
(I added a JUnit test to the example project, that makes sure, that all files get generated.
Should I pull it into a separate project?)

https://github.com/borisbrodski/Jnario/tree/no_xtend_xtext2.9

**Comment by borisbrodski** [comment]
Jnario working good with Xtext 2.9 by now. Do we need the 'Xtext 2.8.x' version of Jnario?

If no, I would push the current 'Xtext 2.9.x'  version of Jnario to our no_xtend branch and continue from there.

**Comment by ghaith** [comment]
TLDR: I have some rich string fixes but i will port them to 2.9

I have some edits on a 2.8 based version that I have not yet pushed to your branch, they are basically a rework of rich strings because of some inconsitencies we got running our tests. The state can be found in my branch called usable_no_xtend I hadn't had the chance to clean them up and create a pull request, what's missing there is to actually make this change on a higher level and not just for specs. Also the shortcuts for adding placeholders is still not enabled. The implementation itself is close to what xtend did, it converts the rich string segment into a StringConcatination but is still missing some of the smart indentation features that xtend has.
I will be eventually working on porting these changes to 2.9 so there is no need to pull these changes yet into the 2.8 branch, nor would I recommend it as it only works with specs as mentioned above.

**Comment by borisbrodski** [comment]
This sounds great! I leave both branches for a while. My new commits I will push to the `no_xtend_xtext2.9` though.

**Comment by borisbrodski** [comment]
Java 8 say:

> '_' should not be used as an identifier, since it is a reserved keyword from source level 1.8 on

We should fix this as well.

**Comment by borisbrodski** [comment]
Hi!

One question: `exampleTable.forEach` doesn't bind to `Each.forEach(ExampleTable<T> table, Procedure1<T> assertion)` any more. This breaks pretty example table exception messages, like this one:

			example failed

			        | value1     | value2     | sum     |
			        | 1          | 2          | 3       | ✓
			        | 4          | 5          | 7       | ✘     (1)
			        | 7          | 8          | 14      | ✘     (2)

Instead it binds to `org.jnario.lib.ExampleTable.forEach`

We register the `Each` class in `JnarioImplicitlyImportedFeatures` as a _static_ and _extension_ class. Any ideas, why this may happen?

Thanks!

**Comment by borisbrodski** [comment]
Ok, the problem is weird :(

In Java 8 the `Iterator` interface defines default method `forEach()`, that gets bound prior to our extension method `Each.forEach()`. The new default method signature is

```java
public interface Iterable<T> {
  default void forEach(Consumer<? super T> action);
}
```

Since Jnario get compiled with Java 6, I can't override this method in `ExampleTable` (No `Consumer` class defined yet).

My solution: override `getBestCandidate()` method (in `JnarioTypeComputer`) letting `Each.forEach` method always win:

```java
	/*
	 * The method {@link Each#forEach(org.jnario.lib.ExampleTable, org.eclipse.xtext.xbase.lib.Procedures.Procedure1)} get shadowed by
	 * Java 8 method <code>Iterable.forEach(...)</code>. This will be overritten here. The our <code>forEach</code> always win.
	 */
	@Override
	protected ILinkingCandidate getBestCandidate(List<? extends ILinkingCandidate> candidates) {
		for (ILinkingCandidate candidate : candidates) {
			if (candidate.getFeature() instanceof JvmOperation && candidate.getFeature().eContainer() instanceof JvmGenericType) {
                JvmOperation feature = (JvmOperation) candidate.getFeature();
                JvmGenericType type = (JvmGenericType) feature.eContainer();
                if (Each.class.getName().equals(type.getQualifiedName()) && "forEach".equals(feature.getSimpleName())) {
                    return candidate;
                }
			}
		}
		return super.getBestCandidate(candidates);
	}
```

Any objections to this solution? (Silence will be considered as "great job!") :))))

**Comment by oehme** [comment]
Why not override the forEach method in `ExampleTable` to execute the logic you want?

**Comment by borisbrodski** [comment]
Because the class `Consumer` is not on the classpath :(

**Comment by oehme** [comment]
You can accept `Object` and do a little reflection magic ;)

**Comment by borisbrodski** [comment]
Unfortunately it didn't worked out:
![image](https://cloud.githubusercontent.com/assets/157821/15427536/136e6fae-1e96-11e6-80f0-424cb1d0689e.png)

Xbase still bounds to `iterator.forEach`:
![image](https://cloud.githubusercontent.com/assets/157821/15427562/41e361e6-1e96-11e6-83c2-77fe9d57176c.png)

Changing method signature to `public <T> void forEach(Procedure1<T> prod)` I get:
![image](https://cloud.githubusercontent.com/assets/157821/15427641/d0460862-1e96-11e6-9a42-c260328e296e.png)

So I see no other way, but to mess up with the TypeComputer...

**Comment by borisbrodski** [comment]
I would like to automate release process a bit. Is there a particular reason, why we gave 3 different groupIds in the project:

* org.jnario
* org.jnario.feature
* org.jnario.plugin

The problem is, that the tycho set-version plugin only supports changing versions within a single groupId (see artifacts attribute). https://eclipse.org/tycho/sitedocs/tycho-release/tycho-versions-plugin/set-version-mojo.html

Should I change this?

**Comment by SebastianPoetzsch** [comment]
#[1,2,3] => [
        	forEach [println(it)]
        ]

not working after override `forEach` in `getBestCandidate()` method

**Comment by borisbrodski** [comment]
Considering set-version post from above:

I will write XSLT and run it use `org.codehaus.mojo:xml-maven-plugin:transform` like this

```
<configuration>
  <transformationSets>
    <transformationSet>
      <dir>.</dir>
      <outputDir>.</outputDir>
      <includes>
        <include>pom.xml</include>
        <include>plugins/org.jnario/pom.xml</include>
        <include>plugins/org.jnario.lib/pom.xml</include>
        ...
      </includes>
      <stylesheet>set-version-override-parent-version.xslt</stylesheet>
      <parameters>
        <parameter>
          <name>VERSION</name>
          <value>${VERSION}</value>
        </parameter>
      </parameters>
    </transformationSet>
  </transformationSets>
</configuration>
```

This will allow us to change version of the Jnario with a single call

```
mvn org.codehaus.mojo:xml-maven-plugin:transform "-DVERSION=$1"
```

Any objections to this?

**Comment by riederm** [comment]
I created a branch where I added a common base-Grammar for the Spec and the Feature-grammars in Jnario. It is basically Xbase with the Richstrings (Template-Expressions) from Xtend. I called it **XbaseWithRichstrings**. I also ported all Richstring-related code from Xtend to this grammar-plugin. https://github.com/riederm/Jnario/tree/xbaseWithRichstrings_2.11

So this means that Specs and Features can use Richstrings exactly like we were used to them when Jnario was based on Xtend. Following features are supported:

- Syntax Highlighting (including syntax highlighting of the resulting indentation)
- Template expressions
- FOR-loops inside Richstrings  including BEFORE, SEPARATOR, AFTER
- IF inside Richstrings
- AutoEdit strategy that handles Richstrings correctly

The only thing that does not work atm are Comments inside Richstrings («« comment »») - but I guess this will not take too long to add this feature as well.
I did this because we use richstrings quite intensively in our company, and its really a pain without the proper syntax-highlighting. Sometimes I also feel that the indentation behaviour did not exactly match the Xtend behavior. Also the suggestions form @oehme, to use Xtend-files in combination with specs and features feels somehow disruptive (at least for me?).

I also moved the Xtext-version to **Xtext 2.11**. This means that I also moved the Java versions to **Java 1.8**. So the Branch has a lot of changes, I regenerated the grammars and the version-switch obviously means that basically all generated files look somehow different, so really a ton of changes in terms of git :-(.  I pulled the last changes from @borisbrodski 's Xtext2.9 branch and it also successfully builds on travis (https://travis-ci.org/riederm/Jnario/builds/221095372). 

So @borisbrodski what do you think? You feel like integrating this stuff into your latest branch and eventually we gonna move everything finally to master? I know that @oehme suggested to drop the Xtend-Template-Expressions from Jnario, I find it not really satisfying. I also think that maybe what I did for the XbaseWithRichstrings-Grammar could eventually be merged into the original Xbase-grammar since I guess Jnario could show that Richstrings successfully work in pure xbase languages?

Whats the proper workflow here? since i also did a version-jump it feels strange to merge it back into the Xtext.29 branch?

**Comment by borisbrodski** [comment]
That is fantastic!

The Xtend-RichStrings is definitely one of the key Xtend- or actually Xbase-features.
Great to have it in Jnario again.

I will build and test it with our "big" project.

**Comment by borisbrodski** [comment]
Just pushed most resent version of Jnario https://github.com/borisbrodski/Jnario/tree/master incorporating work of 

* Mathias Rieder
* Ghaith Hachem
* Sebastian Pötzsch
* Dominik Gabriel

(sorry, if some was not mentioned)

and adding following 

* Migration to Xtext 2.14
* Migration from PegDown to FlexMark
* Improving performance (see `Should: let equals() do the type checking` commit)
* Fixing some bugs
  * Fix "Organize Imports including Jnario"
  * Fix "Open generated file"
  * Fix Guillemet (Ctrl+<, Ctrl+Shift+<)

HUGE thank you to all of you!

_Please, if possible, test the last Jnario with your projects!_

I plan to create a new issue, like "Make Jnario great again", in order to discuss further steps towards Jnario 2.0 release!
